src/tiled/mapscene.cpp

```diff
@@ -166,6 +166,33 @@ void MapScene::setMapDocument(MapDocument *mapDocument)
     refreshScene();
 }
 
+// If none of selected objects is on current layer, sets current layer to one with selected objects
+static void updateCurrentLayerBySelection(MapDocument *mapDocument)
+{
+    QList<MapObject*> selectedObjects = mapDocument->selectedObjects();
+
+    if (selectedObjects.empty())
+        return;
+
+    ObjectGroup *currentObjectLayer = dynamic_cast<ObjectGroup *>(mapDocument->currentLayer());
+
+    // Return if any of selected objects is in selected layer
+    for (MapObject *object : selectedObjects) {
+        if (currentObjectLayer->objects().contains(object))
+            return;
+    }
+
+    // Find layer that contains first selected object, and set it as current
+    for (Layer *layer : mapDocument->map()->layers()) {
```

Comment:
This unfortunately no longer iterates all layers. You need to use the `LayerIterator` for that, because the layers for a hierarchy. However, you actually don't need to iterate all layers, because you're looking for `selectedObjects.first()->objectGroup()`.